Hashkey integration

rayBastard opened the pull request.

rayBastard committed:
hashkey.ts created

rayBastard committed:
abstract/hashkey.ts added

yzh-pelle committed:
fetchMarkets added

yzh-pelle committed:
Update abstract/hashkey.ts

rayBastard committed:
fees added

rayBastard committed:
parseMarket updated

rayBastard committed:
parseMarket updated

rayBastard committed:
fetchOrderBook added

rayBastard committed:
fetchTrades added

rayBastard committed:
fetchOHLCV added

rayBastard committed:
fetchTicker added

rayBastard committed:
parseTrade updated

rayBastard committed:
sign updated and fetchBalance added

rayBastard committed:
fetchBalance updated

rayBastard committed:
Networks added

rayBastard committed:
Update hashkey.ts

yzh-pelle committed:
fetchMarkets updated and fetchCurrencies added

yzh-pelle committed:
networks, markets and currencies updated

yzh-pelle committed:
Update skip-tests.json

yzh-pelle committed:
a bit of updates

yzh-pelle committed:
Update skip-tests.json

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
networks updated

rayBastard committed:
static/currencies/hashkey.json added

rayBastard committed:
static/markets/hashkey.json added

rayBastard committed:
Update hashkey.ts

rayBastard committed:
static/request/hashkey.json added

rayBastard committed:
static/response/hashkey.json added

rayBastard committed:
request/hashkey.json updated

rayBastard committed:
currencies/hashkey.json fixed

rayBastard committed:
response/hashkey.json updated

rayBastard committed:
request/hashkey.json updated

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Update response/hashkey.json

rayBastard committed:
Merge branch 'ccxt:master' into hashkey-integration

yzh-pelle committed:
fetchLastPrices added

yzh-pelle committed:
sign updated

rayBastard committed:
Update response/hashkey.json

rayBastard committed:
defaultNetworkCode fixed

rayBastard committed:
fetchDepositAddress added

rayBastard committed:
request/hashkey.json updated

rayBastard committed:
request/hashkey.json

rayBastard committed:
response/hashkey.json

yzh-pelle committed:
createOrder for spot markets added

rayBastard committed:
Update abstract/hashkey.ts

rayBastard committed:
fetchDeposits added

rayBastard committed:
hashkey.ts updated

rayBastard committed:
Update hashkey.ts

rayBastard committed:
fetchWithdrawals added

rayBastard committed:
Update hashkey.ts

rayBastard committed:
createSpotOrder fixed

rayBastard committed:
createSpotOrder fixed

1alexbc1 reviewed (approved):
Proof

yzh-pelle committed:
Update request/hashkey.json

yzh-pelle committed:
Update response/hashkey.json

yzh-pelle committed:
has.createOrders is set to true

1alexbc1 reviewed (approved):
Revolut

yzh-pelle committed:
fetchDiposits/Withdrawals updated

1alexbc1 reviewed (approved):
Proof

yzh-pelle committed:
fetchOrder added

1alexbc1 reviewed (approved):
Proof

yzh-pelle committed:
Update response/hashkey.json

yzh-pelle committed:
handleOptionAndParams are changed to handleOption

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
cancelOrder for spot added

1alexbc1 reviewed (approved):
Proof

yzh-pelle committed:
fetchOpenOrders for spot added

yzh-pelle committed:
fetchCanceledAndClosedOrders added

rayBastard committed:
cancelAllOrders added

rayBastard committed:
cancelOrders added

rayBastard committed:
cancelOrders updated

rayBastard committed:
customUrlencode added

rayBastard committed:
fetchAccounts added

rayBastard committed:
fetchMyTrades added

rayBastard committed:
transfer added

rayBastard committed:
response/hashkey.json updated

rayBastard committed:
abstract/hashkey.ts updated

yzh-pelle committed:
fetchFundingRates added

yzh-pelle committed:
fetchFundingRateHistory added

yzh-pelle committed:
Update hashkey.ts

1alexbc1 reviewed (approved):
Proof

1alexbc1 reviewed (commented):
Proof

1alexbc1 reviewed (commented):
Proof

rayBastard committed:
withdraw added

yzh-pelle committed:
Update abstract/hashkey.ts

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
fetchFundingRate added

yzh-pelle committed:
fetchTickers added

yzh-pelle committed:
fetchLeverageTiers added

yzh-pelle committed:
fetchTradingFee/Fees added

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
parseMarket updated

yzh-pelle committed:
createOrder and parseOrder updated for swap markets

yzh-pelle committed:
fetchPositions added

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Merge branch 'ccxt:master' into hashkey-integration

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
fetchLedger added

rayBastard committed:
Update skip-tests.json

rayBastard committed:
Merge branch 'master' into hashkey-integration

yzh-pelle committed:
fetchOrder updated for swap markets

yzh-pelle committed:
fetchOpenOrders updated for swap markets

yzh-pelle committed:
fetchCanceledAndClosedOrders updated for swap markets

yzh-pelle committed:
fetchMyTrades updated for swap markets

yzh-pelle committed:
cancelOrder updated for swap markets

yzh-pelle committed:
cancelAllOrders updated for swap markets

yzh-pelle committed:
cancelOrders updated for swap markets

rayBastard committed:
fetchLeverage added

rayBastard committed:
setLeverage added

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
createOrders added

rayBastard committed:
fees updated for swap

rayBastard committed:
fetchAccounts updated

rayBastard committed:
pro/hashkey.ts added

rayBastard committed:
Update pro/hashkey.ts

rayBastard committed:
Update pro/hashkey.ts

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Update response/hashkey.json

yzh-pelle committed:
watch OHLCV added

yzh-pelle committed:
parseTransfer updated

yzh-pelle committed:
watchTicker added

yzh-pelle committed:
Update pro/hashkey.ts

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Update response/hashkey.json

yzh-pelle committed:
parseMarket and fees updated

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Update request/hashkey.json

yzh-pelle committed:
watchTrades added

yzh-pelle committed:
watchOrderBook added

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Merge branch 'ccxt:master' into hashkey-integration

rayBastard committed:
fetchMyTrades fixed

rayBastard committed:
fetchMyTrades, fetchOpenOrders, fetchCanceledAndClosedOrders updated with sub account endpoints for swap

rayBastard committed:
handleErrors added

yzh-pelle committed:
Update abstract/hashkey.ts

carlosmiei committed:
optional TIF

yzh-pelle committed:
parseOrder updated and additional endpoints added

yzh-pelle committed:
watchOrders added

yzh-pelle committed:
watchMyTrades added

yzh-pelle committed:
parsing of PO updated

yzh-pelle committed:
Update pro/hashkey.ts

rayBastard committed:
Update response/hashkey.json

rayBastard committed:
Update response/hashkey.json

yzh-pelle committed:
watchPositions added

rayBastard committed:
fetchOpenOrders updated

yzh-pelle committed:
Update response/hashkey.json

yzh-pelle committed:
fetchCanceledAndClosedOrders updated

yzh-pelle committed:
handleErrors updated

yzh-pelle committed:
handleErrors and fetchBalance updated

yzh-pelle committed:
endpoints for listenKey updated

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
watchBalance added

yzh-pelle committed:
some `//todo` removed

rayBastard committed:
createMarketBuyOrderWithCost updated

rayBastard committed:
parseMarket updated

rayBastard committed:
parseMarket updated

rayBastard committed:
handleOrderBook updated

rayBastard committed:
Update skip-tests.json

yzh-pelle committed:
trying to get array length in parseMarket without an error

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+        let minLeverage: Int = undefined;
+        let maxLeverage: Int = undefined;
+        if (isSwap) {
+            amountPrecisionString = Precise.stringDiv (amountPrecisionString, contractSizeString);
```
I think this is wrong, it seems that `stepSize` is the amount precision already you don't need to divide here

yzh-pelle replied:
The documentation of the exchange says that all swap amount of contracts should be an Int. Deleting this division causes that `amountToPresicion` doesn't throw an exception when amount is less than 1, and the exchange returns `{"code":"-1202","msg":"Order sell quantity too small"}`

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+        let maxLeverage: Int = undefined;
+        if (isSwap) {
+            amountPrecisionString = Precise.stringDiv (amountPrecisionString, contractSizeString);
+            amountMinLimitString = Precise.stringDiv (amountMinLimitString, contractSizeString);
```
same with amountMin and amountMax

yzh-pelle replied:
The same as above

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+        }
+        const tradingFees = this.safeDict (this.fees, 'trading');
+        const fees = isSpot ? this.safeDict (tradingFees, 'spot') : this.safeDict (tradingFees, 'swap');
+        return {
```
we should parse this through `safeMarketStructure`

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+            if (market !== undefined) {
+                request['symbol'] = market['id'];
+            }
+            let clientOrderId: Str = undefined;
```
since the unified term `clientOrderId` is the same as the exchange-specific one, we odn't need to do this because if user provides it in `params` we extend it at the end with request

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+        if (until !== undefined) {
+            request['endTime'] = until;
+        }
+        let fromId: Str = undefined;
+        [ fromId, params ] = this.handleOptionAndParams (params, methodName, 'fromId');
+        if (fromId !== undefined) {
+            request['fromId'] = fromId;
+        }
+        let toId: Str = undefined;
+        [ toId, params ] = this.handleOptionAndParams (params, methodName, 'toId');
+        if (toId !== undefined) {
+            request['toId'] = toId;
+        }
+        let accountId: Str = undefined;
+        [ accountId, params ] = this.handleOptionAndParams (params, methodName, 'accountId');
```
same here, we should keep them in the docs, but we don't need this.

yzh-pelle replied:
The same - just to search 'preinstalled' values in `options`

carlosmiei replied:
but in this case it is not needed because "fromId/toId is not a "fixed" value, users have to adjust it frequently so it's not a value to store in options

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+         */
+        const methodName = 'fetchOHLCV';
+        await this.loadMarkets ();
+        const market = this.market (symbol);
```
can we add automatic pagination here? check for instance how bybit does it (it's easy)

rayBastard replied:
Done

yzh-pelle committed:
update after review

https://github.com/ccxt/ccxt/pull/23176#discussion_r1711306671

rayBastard committed:
fetchOHLCV updated with automatic pagination

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+        const request: Dict = {};
+        let clientOrderId: Str = undefined;
+        [ clientOrderId, params ] = this.handleParamString (params, 'clientOrderId');
+        if (id !== undefined) {
```
due to the method's signature, `id` is always expected to be defined, so we should first check if clientOrderid is not undefined, and then use the id

rayBastard replied:
done

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4336 @@
+
+    parseLeverage (leverage: Dict, market: Market = undefined): Leverage {
+        let marginMode = this.safeString (leverage, 'marginType');
+        if (marginMode !== undefined) {
```
we can use safeStringLower here instead

rayBastard replied:
Sure, sometimes forget about useful methods

carlosmiei commented on ts/src/pro/hashkey.ts:
```diff
@@ -0,0 +1,852 @@
+         * @param {object} [params] extra parameters specific to the exchange API endpoint
+         * @returns {object[]} a list of [order structures]{@link https://docs.ccxt.com/#/?id=order-structure}
+         */
+        await this.loadMarkets ();
```
will this work for both spot and swap orders?

rayBastard replied:
Let us check

yzh-pelle replied:
Yes it works for both types of markets

rayBastard committed:
fetchOrder updated

rayBastard committed:
parseLeverage fixed

rayBastard committed:
Update response/hashkey.json

rayBastard committed:
Update response/hashkey.json

yzh-pelle committed:
fromId and toId removed from methods

https://github.com/ccxt/ccxt/pull/23176#discussion_r1711441907

yzh-pelle committed:
handleBalance update

yzh-pelle committed:
Update response/hashkey.json

yzh-pelle committed:
parseOrder and handleErrors updated

yzh-pelle committed:
Update response/hashkey.json

yzh-pelle committed:
parseOrder updated

rayBastard committed:
parseTrade updated

rayBastard committed:
Update response/hashkey.json

rayBastard committed:
parseTrade updated

rayBastard committed:
Update response/hashkey.json

carlosmiei committed:
Merge branch 'master' of github.com:ccxt/ccxt into hashkey-integration

carlosmiei committed:
fix static tests

rayBastard committed:
Update hashkey.ts

yzh-pelle committed:
Some `//todo` removed

yzh-pelle committed:
createOrder response tests for swap deleted

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        }
+        const clientOrderId = this.safeString (params, 'clientOrderId');
+        if (clientOrderId === undefined) {
+            throw new ArgumentsRequired (this.id + ' createSwapOrder() requires a clientOrderId parameter');
```
if the user does not provide it we should generate one using for instance this.uuid instead of throwing.


Besides that, since `createSwapOrder` is not a unified method, if we need to throw an exception, we should use `createOrder` as the method's name instead

yzh-pelle replied:
Thanks, will know!

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        if (clientOrderId === undefined) {
+            throw new ArgumentsRequired (this.id + ' createSwapOrder() requires a clientOrderId parameter');
+        }
+        const triggerPrice = this.safeNumber (params, 'triggerPrice');
```
we should never use safeNumber to build up the request because its return type is dynamic, since we're going to format it we can use safeString here

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        this.checkTypeParam (methodName, params);
+        await this.loadMarkets ();
+        const request: Dict = {};
+        if (id !== undefined) {
+            request['orderId'] = id;
+        } else {
+            const clientOrderId = this.safeString (params, 'clientOrderId');
+            if (clientOrderId === undefined) {
+                throw new ArgumentsRequired (this.id + ' ' + methodName + '() requires an id argument or clientOrderId parameter');
+            }
+        }
```
we should do the opposite here, first check if cancelOrderId is available and use it, otherwise default to id, because `id` is expected to be defined according to the signature so users can provide smth like `id="random"` and `clientOrderId` in `params`

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+            'symbol': market['id'],
+        };
+        let side: Str = undefined;
+        [ side, params ] = this.handleOptionAndParams (params, 'cancelAllOrders', 'side');
```
again this information is very dynamic so I don't think we should use handleOptionAndParams, just let the user provide it in `params` if needed

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+            request['side'] = side;
+        }
+        if (market['spot']) {
+            return await this.privateDeleteApiV1SpotOpenOrders (this.extend (request, params));
```
cancelAllOrders should return the unified type `Order[]` (check unified implementations)

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        }
+    }
+
+    async cancelOrders (ids: string[], symbol: Str = undefined, params = {}) {
```
same thing here, we should return `Order[]`

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        await this.loadMarkets ();
+        const request = {};
+        let orderIds = '';
+        for (let i = 0; i < ids.length; i++) {
+            orderIds += this.safeString (ids, i) + ',';
+        }
+        orderIds = orderIds.slice (0, -1);
```
can't this be replaced with only `ordersIds.join(',')` ?

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        if (clientOrderId !== undefined) {
+            request['origClientOrderId'] = clientOrderId;
+        } else if (id === undefined) {
+            throw new ArgumentsRequired (this.id + ' ' + methodName + '() requires an id argument or clientOrderId parameter');
```
in this method you can assume `id` is defined

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+        [ marketType, params ] = this.handleMarketTypeAndParams (methodName, market, params, marketType);
+        let response = undefined;
+        if (marketType === 'spot') {
+            if (id !== undefined) {
```
same thing here

carlosmiei commented on ts/src/hashkey.ts:
```diff
@@ -0,0 +1,4313 @@
+         * @returns {object[]} a list of [position structure]{@link https://docs.ccxt.com/#/?id=position-structure}
+         */
+        const methodName = 'fetchPositions';
+        if ((symbols === undefined) || (symbols.length < 1)) {
```
`symbols.length` should be in an isolated statement like `const symbolsLength = symbols.length` otherwise will transpile as str_len in php

yzh-pelle committed:
response and request createOrder for swap deleted

yzh-pelle committed:
watchTrades - trades sorted in ascending direction

yzh-pelle committed:
handleTrades updated

yzh-pelle committed:
Handled with `// todo` at hashkey.ts

yzh-pelle committed:
Mandatory clientOrderId are generated by `this.uuid()`

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715016842

yzh-pelle committed:
createSwapOrder updated

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715017571

yzh-pelle committed:
cancelOrder updated

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715019775

yzh-pelle committed:
cancelAllOrders updated

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715020955

yzh-pelle committed:
cancelAllOrders updated

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715021539

yzh-pelle committed:
cancelOrders updated

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715022071

yzh-pelle committed:
Update response/hashkey.json

yzh-pelle committed:
Update request/hashkey.json

yzh-pelle committed:
Update hashkey.ts

yzh-pelle committed:
Update hashkey.ts

https://github.com/ccxt/ccxt/pull/23176#discussion_r1715029609
https://github.com/ccxt/ccxt/pull/23176#discussion_r1715025960
https://github.com/ccxt/ccxt/pull/23176#discussion_r1715025820
https://github.com/ccxt/ccxt/pull/23176#discussion_r1715023757

carlosmiei committed:
add broker-id

carlosmiei committed:
fix fetchPositions condition

carlosmiei committed:
update error codes

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Update response/hashkey.json

rayBastard committed:
Update request/hashkey.json

rayBastard committed:
Update response/hashkey.json

carlosmiei committed:
Merge branch 'hashkey-integration' of github.com:rayBastard/ccxt into hashkey-integration

carlosmiei committed:
add logo

carlosmiei commented:
![1x](https://github.com/user-attachments/assets/6dd6127b-cc19-4a13-9b29-a98d81f80e98)

carlosmiei committed:
update input source
